docs/design/upgrade-resource-state.md

```diff
@@ -0,0 +1,581 @@
+
+For example in the framework:
+
+```go
+// New type
+type ResourceWithUpgradeState interface {
+    UpgradeState(context.Context, /*...*/) /*...*/
+}
+```
+
+Which would result in the following optional provider implementation:
+
+```go
+func (r ExampleResource) UpgradeState(ctx context.Context, /*...*/) /*...*/ {
+    /* ... */
+}
+```
+
+There are considerations for the possible method parameters and returns, which is why they are omitted above and they will be discussed below.
+
+This would ensure framework resources:
+
+- Have access to the provider client during upgrade state operations.
+- Are simpler in implementation until the additional complexity is necessary.
+
+However, there are some drawbacks:
+
+- If `Schema.Version` is greater than `0`, there are no compile time errors if the `UpgradeState` method is not defined. Framework-defined unit testing may be possible to ensure that if the `Schema.Version` returned from `(ResourceType).GetSchema()` is greater than `0`, that `(ResourceType).NewResource()` returns a type that supports the `ResourceWithUpgradeState` interface, however it would require provider developers to ensure they implement the additional unit testing.
+
+#### Direct Request and Response Parameters
+
+The framework could require provider developers to implement the `UpgradeState` method signature similar to other `Resource` functionality, such as `Create`, `Read`, etc.
+
+For example in the framework:
+
+```go
+// New type
+type ResourceWithUpgradeState interface {
+    UpgradeState(context.Context, UpgradeResourceStateRequest, *UpgradeResourceStateResponse)
+}
+
+// New type
+type UpgradeResourceStateRequest struct {
+    // Current state version
+    // May be worth renaming to CurrentStateVersion in the actual implementation
+    Version int64
+
+    // JSON encoded or flatmap state
+    // The type would likely be abstracted in the framework, but showing for brevity.
+    RawState tfprotov6.RawState
+}
+
+// New type
+type UpgradeResourceStateResponse struct {
+    // Upgraded state, which must be sent as msgpack
+    // The underlying *tftypes.DynamicValue type is very unfriendly for
+    // provider developers at this layer so a next option is to rely on
+    // the existing State type.
+    State State
+
+    Diagnostics diag.Diagnostics
+}
+```
+
+Which would result in the following optional provider implementation:
+
+```go
+func (r ExampleResource) UpgradeState(context.Context, req tfsdk.UpgradeResourceStateRequest, *tfsdk.UpgradeResourceStateResponse) {
+    // potentially branching logic based on req.Version
+    // complex logic to convert req.RawState into the resp.State
+}
+```
+
+In this setup, the framework could only implement a thin abstraction over the request data since it has no previous schema information. One framework workaround to remove this hurdle for provider developers might be to introduce additional method on `ResourceType` or `Resource` to fetch previous schema.
```

Comment:
introduce additional method -> introduce an additional method

Reply:
```suggestion
In this setup, the framework could only implement a thin abstraction over the request data since it has no previous schema information. One framework workaround to remove this hurdle for provider developers might be to introduce an additional method on `ResourceType` or `Resource` to fetch previous schema.
```